### PR TITLE
build(webpack): Upgrade to Webpack 5 to align with the upstream

### PR DESCRIPTION
**Pull request by azatsarynnyy**
Signed-off-by: Artem Zatsarynnyi <azatsary@redhat.com>

### What does this PR do?
Adapts Che-Theia to using Webpack 5 to align with the upstream Theia.

### What issues does this PR fix or reference?

fixes https://github.com/eclipse/che/issues/19804

### How to test this PR?

**Without CDN support**
1. Use the Devfile posted by che-bot.
2. Check that the following files are loaded from the Express server (Che-Theia backend):
image

**With CDN support**
1. Use the following Devfile:
```yaml
apiVersion: 1.0.0
metadata:
  name: with-cdn
attributes:
  persistVolumes: 'false'
components:
  - memoryLimit: 512Mi
    type: cheEditor
    reference: 'https://gist.githubusercontent.com/azatsarynnyy/09d57d7596a44d644d1774c2fe2a2b01/raw/ccd03e4a6dd30c724778ad70af70f4d9f6d2e6d8/webpack-cdn'
```
2. Check that the following files are loaded from CDN:
image

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable

### CODE REVIEW

File: build.include [hunk 1]
```diff
@@ -14,7 +14,7 @@ IMAGE_TAG="next"
 THEIA_GITHUB_REPO="eclipse-theia/theia"
 THEIA_VERSION="master"
 THEIA_BRANCH="master"
-THEIA_COMMIT_SHA="2eb4319fb459f30ad897e8467878154c6790aaef"
+THEIA_COMMIT_SHA="d734315e505ad4a7691a3f6f4a07d2209af0d9cf"
```
Comment (azatsarynnyy) [inline build.include:17]: further adaptation will be done in https://github.com/eclipse-che/che-theia/pull/1124

File: extensions/eclipse-che-theia-plugin-ext/package.json [hunk 2]
```diff
@@ -25,7 +25,8 @@
     "js-yaml": "3.13.1",
     "drivelist": "9.0.2",
     "@eclipse-che/theia-remote-api": "^0.0.1",
-    "@eclipse-che/workspace-telemetry-client": "latest"
+    "@eclipse-che/workspace-telemetry-client": "latest",
+    "mime": "2.5.2"
```
Comment (azatsarynnyy) [inline extensions/eclipse-che-theia-plugin-ext/package.json:29]: I'll create a CQ for `mime`.

File: generator/package.json [hunk 3]
```diff
@@ -15,10 +15,11 @@
...
     "read-pkg": "3.0.0",
     "yargs": "12.0.5",
-    "webpack": "^4.0.0",
-    "html-webpack-plugin": "^3.2.0",
+    "webpack": "^5.36.2",
+    "html-webpack-plugin": "^5.3.1",
     "axios": "0.21.1",
-    "tmp": "^0.0.33"
+    "tmp": "^0.0.33",
+    "process": "^0.11.10"
```
Comment (azatsarynnyy) [inline generator/package.json:22]: I'll create a CQ for `process`.

**Review by benoitf — APPROVED** [review]
image

image

with cdn on I had a failure on recommendation plugin about a missing to-regex package

image

File: extensions/eclipse-che-theia-plugin-ext/package.json (hunk 2 above)
Comment (azatsarynnyy) [reply extensions/eclipse-che-theia-plugin-ext/package.json:29]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=23451

File: generator/package.json (hunk 3 above)
Comment (azatsarynnyy) [reply generator/package.json:22]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=23452

**Comment by azatsarynnyy** [comment]
Thanks! I'll check it.

File: generator/package.json [hunk 4]
```diff
@@ -15,10 +15,11 @@
     "js-yaml": "3.13.1",
     "read-pkg": "3.0.0",
     "yargs": "12.0.5",
-    "webpack": "^4.0.0",
-    "html-webpack-plugin": "^3.2.0",
+    "webpack": "^5.36.2",
```
Comment (tsmaeder) [inline generator/package.json:18]: I'm wondering why these are not devDependencies.

Comment (azatsarynnyy) [reply generator/package.json:18]: For che-theia generator they are actually the production dependencies. As they used when we run the generator binary on the Theia sources.

**Comment by azatsarynnyy** [comment]
hmm... added `to-regex`, but now it complains to `array-unique`
image

File: generator/package.json (hunk 4 above)
Comment (tsmaeder) [reply generator/package.json:18]: Hmh...isn't the webpack config executed as part of running webpack and therefore the dependencies would be satisfied from whatever project webpack is used in? So shouldn't the "assembly" bring the webpack plugin as devDependencies?

File: plugins/recommendations-plugin/package.json [hunk 5]
```diff
@@ -19,7 +19,8 @@
     "globby": "^11.0.1",
     "inversify": "^5.0.1",
     "kind-of": "^6.0.3",
-    "reflect-metadata": "^0.1.13"
+    "reflect-metadata": "^0.1.13",
```
Comment (tsmaeder) [inline plugins/recommendations-plugin/package.json:22]: I'm wondering how this is related to webpack 4 -> webpack 5. Maybe it's an independent problem and should be left for another PR.

File: generator/src/cdn/webpack.config.ts [hunk 6]
```diff
@@ -10,30 +10,11 @@
 
 import { customizeWebpackConfig } from './webpack-customizer';
 
-const yargs = require('yargs');
```
Comment (tsmaeder) [inline generator/src/cdn/webpack.config.ts:13]: Is this change necessary or just a "drive-by" fix?

File: generator/tests/cdn/webpack-customizer.spec.ts [hunk 7]
```diff
@@ -87,7 +89,7 @@ describe('Test webpack customizer', () => {
         expect(baseConfig.entry['cdn-support']).toEqual(path.resolve(__dirname, '../../src/cdn/bootstrap.js'));
         expect(baseConfig.output.filename).toBe('[name].[chunkhash].js');
         expect(baseConfig.optimization.runtimeChunk).toBe('single');
-        expect(baseConfig.optimization.splitChunks.cacheGroups.vendors).toEqual({
+        expect(baseConfig.optimization.splitChunks.cacheGroups.defaultVendors).toEqual({
```
Comment (tsmaeder) [inline generator/tests/cdn/webpack-customizer.spec.ts:92]: Is this a functional change or just a rename to make the code clearer?

File: generator/src/cdn/webpack-customizer.ts [hunk 8]
```diff
@@ -102,11 +109,6 @@ export function customizeWebpackConfig(
             })
         );
 
-        // Use hashed module IDs to ease caching support
-        // and avoid the hash-based chunk names being changed
-        // unexpectedly
-        baseConfig.plugins.push(new webpack.HashedModuleIdsPlugin());
-
         // Insert a custom loader to override file and url loaders,
```
Comment (tsmaeder) [inline generator/src/cdn/webpack-customizer.ts:112]: Since the newly generated webpack config does not use url-loader/file-loader anymore, how is the replacement of the loader with the cdn/webpack-loader.js done? I suspect this would not work when deployed to a cdn.

File: generator/tests/cdn/webpack-customizer.spec.ts (hunk 7 above)
Comment (azatsarynnyy) [reply generator/tests/cdn/webpack-customizer.spec.ts:92]: It's a part of v4->v5 migration https://webpack.js.org/migrate/5/#update-outdated-options
See `optimization.splitChunks.cacheGroups.vendors → optimization.splitChunks.cacheGroups.defaultVendors`

File: plugins/recommendations-plugin/package.json (hunk 5 above)
Comment (azatsarynnyy) [reply plugins/recommendations-plugin/package.json:22]: I'm trying to understand it either.
The problem https://github.com/eclipse-che/che-theia/pull/1134#pullrequestreview-681505810 is not reproduced on Che-Theia:next but on this PR.

File: generator/src/cdn/webpack.config.ts (hunk 6 above)
Comment (azatsarynnyy) [reply generator/src/cdn/webpack.config.ts:13]: It's just a dead code, as I understand.
I don't see how it's related to this Webpack config and who calls it.

File: generator/src/cdn/webpack-customizer.ts (hunk 8 above)
Comment (azatsarynnyy) [reply generator/src/cdn/webpack-customizer.ts:112]: Webpack's 4 `HashedModuleIdsPlugin` corresponds to Webpack's 5 `optimization.moduleIds: 'deterministic'`
https://webpack.js.org/migrate/5/#update-outdated-options
See `HashedModuleIdsPlugin → optimization.moduleIds: 'deterministic'`

And `optimization.moduleIds: 'deterministic'` is default now.
https://webpack.js.org/configuration/optimization/#optimizationmoduleids

That means it's not needed anymore.
BTW, with Webpack 5 it fails with:
```
[webpack-cli] TypeError: webpack.HashedModuleIdsPlugin is not a constructor
    at Object.customizeWebpackConfig (/home/theia-dev/theia-source-code/examples/assembly/cdn/webpack-customizer.js:96:33)
    at module.exports (/home/theia-dev/theia-source-code/examples/assembly/cdn/webpack.config.js:17:26)
    at /home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1391:43
    at Array.map (<anonymous>)
    at evaluateConfig (/home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1383:24)
    at /home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1414:53
    at async Promise.all (index 0)
    at async WebpackCLI.resolveConfig (/home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1413:38)
    at async WebpackCLI.createCompiler (/home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1837:22)
    at async WebpackCLI.buildCommand (/home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1952:20)
    at async Command.<anonymous> (/home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:742:25)
    at async Promise.all (index 1)
    at async Command.<anonymous> (/home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1289:13)

Error: webpack exited with an unexpected code: 2.
    at ChildProcess.<anonymous> (/home/theia-dev/theia-source-code/dev-packages/application-manager/lib/application-process.js:67:28)
    at ChildProcess.emit (events.js:314:20)
    at maybeClose (internal/child_process.js:1022:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:287:5)
error Command failed with exit code 1.
```

File: generator/src/cdn/webpack.config.ts (hunk 6 above)
Comment (tsmaeder) [reply generator/src/cdn/webpack.config.ts:13]: Well, because you removed the options, right. The question is was there a reason to remove the options.

File: generator/src/cdn/webpack-customizer.ts (hunk 8 above)
Comment (tsmaeder) [reply generator/src/cdn/webpack-customizer.ts:112]: I was not talking about the hashed module id's, but about the insertion of `cdn/webpack-loader.js` We're doing some pattern matching which will never execute. I can't comment on code that has not changed (stupid github)

File: generator/package.json (hunk 4 above)
Comment (azatsarynnyy) [reply generator/package.json:18]: Probably, it's a good idea to experiment with in one of the H'n'H days. :slightly_smiling_face:

File: generator/src/cdn/webpack.config.ts (hunk 6 above)
Comment (azatsarynnyy) [reply generator/src/cdn/webpack.config.ts:13]: ahh.. Got it. I'm reading the options passed to Webpack through the `env` argument of the exported function.
yargs is not needed here anymore, so typescript complains about not used variables

File: generator/src/cdn/webpack-customizer.ts (hunk 8 above)
Comment (azatsarynnyy) [reply generator/src/cdn/webpack-customizer.ts:112]: I see. I'm not sure yet how this part actually works.

**Comment by azatsarynnyy** [comment]
> hmm... added `to-regex`, but now it complains to `array-unique`
> image

Adding these dependencies fixes the problem, but it's obviously the wrong way to solve it.
I'm still trying to understand why it became necessary :exploding_head:

**Comment by tsmaeder** [comment]
@azatsarynnyy as I understand it, plugins should be self-contained, so you'd have to include anything that is not provided by Theia itself. When you build the recommendations plugin, does it not include a `node_modules` folder with the missing packages included?

**Comment by azatsarynnyy** [comment]
@tsmaeder it includes the `node_modules` folder with the missing packages.
But when I build the recommendations plug-in itself - no missing packages. Everything is included. I suppose some packages might be lost during `che-theia production`.

**Comment by tsmaeder** [comment]
> But when I build the recommendations plug-in itself - no missing packages. Everything is included. I suppose some packages might be lost during che-theia production.

Does `che-theia production` remove things from the built plugin artifacts or rebuild the plugins in a different way? That would be strange, IMO.

**Comment by benoitf** [comment]
@azatsarynnyy I would swap the lines 
https://github.com/eclipse-che/che-theia/blob/master/dockerfiles/theia/Dockerfile#L100-L106

compile plugins before turning che-theia into production

**Comment by tsmaeder** [comment]
@benoitf but the theia plugin packager is supposed to look at the transitive dependencies and complain if one is missing. I switching those lines would work, wouldn't that indicate a bug in either the plugin packager or `che-theia production`?

**Comment by azatsarynnyy** [comment]
> @azatsarynnyy I would swap the lines
> https://github.com/eclipse-che/che-theia/blob/master/dockerfiles/theia/Dockerfile#L100-L106
> 
> compile plugins before turning che-theia into production

@benoitf it didn't help

**Comment by benoitf** [comment]
so when it works it's using braces library `v3.0.2` and when it fails it grabs index.js and package.json of `braces` v2.3.2 (but do not grab to-regex dependency)
I think it's an issue in plugin-packager.


As workaround, as braces is compliant with v2 (and more efficient) we could just use that resolution

**Comment by azatsarynnyy** [comment]
after setting `"braces": "^3.0.2"` in the resolutions, I've got:
image

`braces 3.0.2` doesn't require `to-regex` at all, but `micromatch 3.1.10` depends on `to-regex 3.0.2` which is not present in `node_modules`

I've looked at `che-theia:next`, it includes `braces 2.3.2` that depends on `to-regex 3.0.2` which is present.

So, it seems with this PR we have no `to-regex 3.0.2` regardless of what `braces` version is used: 3.0.2 or 2.3.2

**Comment by benoitf** [comment]
I've opened https://github.com/eclipse-che/che-theia/pull/1137 as a more proper fix anyway

**Comment by benoitf** [comment]
PR for recommendations plug-in merged, so you could avoid any hack in this PR

**Comment by tsmaeder** [comment]
@azatsarynnyy @benoitf while we can probably proceed with merging this PR, I'm a bit dissatisfied that we don't understand what the problem is with the plugin dependencies. IMO, we should try and find a reproducible case and open an issue against the plugin packager, if that is indeed where the bug is. 
In the long run, I think we should just bury the Theia plugin package format and go with the vs code tooling.

**Comment by azatsarynnyy** [comment]
Thanks @benoitf! With your patch included, everything works well now.
Also, now it doesn't require adding `"process": "^0.11.10"` :+1:

**Comment by azatsarynnyy** [comment]
[crw-ci-test]

**Review by benoitf — APPROVED** [review]
image

Works fine

**Comment by benoitf** [comment]
Happy path worked on alpine/next and alpine/stable and ubi8/stable so even if the jenkins one is failing that's not an issue to merge

**Comment by azatsarynnyy** [comment]
> Happy path worked on alpine/next and alpine/stable and ubi8/stable so even if the jenkins one is failing that's not an issue to merge

yeah, but it also depends on the CQ for `mime`
https://github.com/eclipse-che/che-theia/pull/1134/files#diff-242ccd316dba99d9699fc011372a4ddf4d2f6bdf25e9ba4c48f3cbee0aedc66eR29

**Comment by benoitf** [comment]
@azatsarynnyy mime has a score greater than 75 on clearlydefined so I think we can check-in it (self-verification)

> It then sends the content that has not been flagged as either approved or restricted to ClearlyDefined; if an entry is known to ClearlyDefined and has a score of at least 75 and all discovered licenses are on our approved licenses list, then it goes into the “approved” pile.
